File: Javascript/2.Settembre/26.IPv4Validator-5Settembre.js
//todo Data una stringa, determina se si tratta di un indirizzo IPv4 valido. Un indirizzo IPv4 valido è composto da quattro numeri interi separati da punti (.). 
//todo Ogni numero deve soddisfare le seguenti condizioni:
 
//! È compreso tra 0 e 255 inclusi.
//! Non deve contenere zeri iniziali (ad esempio, 0 è consentito, 01 no).
//! Sono consentiti solo caratteri numerici.

//? Tempo di esecuzione:
//* 26 minuti e 23 secondi

//?Codice scritto senza utilizzo di ChatGpt

function isValidIPv4(ipv4) {
  let lastChar = ipv4.slice(-1)
  if(lastChar === ".") return false

  let ipvNumber = ipv4.split(".")
  let ipvNumberArr = ipvNumber.map(Number)

  for(let i = 0; i < 4; i++) {
    if(ipvNumberArr[i] >= 0 && ipvNumberArr[i] < 255) return true
    else return false
  }
}

console.log(isValidIPv4("192.168.101."))

//?Codice scritto da ChatGpt
//* Non utilizzata

//* ✅ Pro

//* Buona idea controllare il punto finale

//* Uso corretto di split(".")

//* Tentativo giusto di convertire in numeri (map(Number))

//* Struttura chiara e leggibile

//! ❌ Contro / Errori

//! return nel for → esci al primo numero (errore grave)

//! Non controlli che ci siano esattamente 4 blocchi

//! 255 escluso (< 255 invece di <= 255)

//! Non gestisci NaN (es. "abc")

//! Accetta numeri vuoti ("192..1.1")

//! Non blocca numeri negativi o con spazi

//? Versione Chat Gpt 

function isValidIPv4(ipv4) {
  const parts = ipv4.split(".");
  if (parts.length !== 4) return false;

  for (let part of parts) {
    if (part === "") return false;
    if (!/^\d+$/.test(part)) return false;

    const num = Number(part);
    if (num < 0 || num > 255) return false;
  }

  return true;
}

console.log(isValidIPv4("192.168.101.1"));   // true
console.log(isValidIPv4("192.168.101."));    // false
console.log(isValidIPv4("256.1.1.1"));       // false


//todo Considerazioni: 

//! N.B.

File: Python/2.Settembre/26.IPv4_validator-5Settembre.py
# # //todo Consegna dell'esercizio

# Data una stringa, determinare se si tratta di un indirizzo IPv4 valido.

# Condizioni:
# - Deve contenere esattamente 4 numeri separati da punti.
# - Ogni numero deve essere compreso tra 0 e 255.
# - Non sono ammessi zeri iniziali (tranne il numero "0").
# - Sono ammessi solo caratteri numerici.

# # //todo

# Bisogna verificare sia la struttura dell'indirizzo sia la validità di ogni singolo blocco.

# # //*Codice scritto senza utilizzo di ChatGpt

def is_valid_ipv4(ipv4):
    ip_split = ipv4.replace(".", " ").split()

    if len(ip_split) < 4:
        return False

    is_ip = True

    for number in ip_split:
        if number.startswith("0") and number != "0":
            return False;

        number = int(number);
        if number < 0 or number > 255:
            return False;

    return is_ip

# # //*

# # //* Pro e contro del mio codice

# # //* Pro

# - Hai gestito correttamente il controllo degli zeri iniziali.
# - Hai verificato il range tra 0 e 255.
# - Hai usato un ciclo semplice e leggibile.
# - Restituisci immediatamente False quando trovi un errore.

# # //! Contro

# - `replace(".", " ").split()` è rischioso: `"192.168.101."` diventa `["192", "168", "101"]`, facendo sparire l'ultimo blocco vuoto.
# - Controlli solo `len(ip_split) < 4`, ma dovresti verificare che sia **esattamente 4**.
# - Se uno dei blocchi contiene lettere (`"192.abc.1.1"`), `int(number)` genera un errore.
# - La variabile `is_ip` è inutile: vale sempre `True` e non viene mai modificata.
# - I `;` a fine riga non servono in Python.

# # //*

# # //* Consigli

# - Usa direttamente:

# ```python
# ip_split = ipv4.split(".")